### PR TITLE
luminous: client: anchor dentries for trimming to make cap traversal safe

### PR DESCRIPTION
**Pull request by batrick**
http://tracker.ceph.com/issues/24185

### CODE REVIEW

**Comment by ukernel** [comment]
/home/zhyan/Ceph/ceph-2/src/client/Client.cc: In member function ‘void Client::trim_caps(MetaSession*, int)’:
/home/zhyan/Ceph/ceph-2/src/client/Client.cc:4104:32: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
   while ((caps_size - trimmed) > max && !p.end()) {

**Comment by batrick** [comment]
@ukernel, thanks PTAL.

File: src/client/Client.cc [hunk 1]
```diff
@@ -4141,8 +4140,11 @@ void Client::trim_caps(MetaSession *s, int max)
...
-  ldout(cct, 20) << " clearing anchored inodes" << dendl;
-  anchor.clear();
+  ldout(cct, 20) << " trimming queued dentries: " << dendl;
+  for (const auto &dn : to_trim) {
+    trim_dentry(dn);
+  }
+  to_trim.clear();
 
   caps_size = s->caps.size();
   if (caps_size > max)
```
Comment (ukernel) [inline src/client/Client.cc]: this line still causes warning. how about change the 'max' parameter to uint64_t

Comment (batrick) [reply src/client/Client.cc]: Thanks, fixed!

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-06-19-1953-luminous](https://trello.com/c/Gd3eYGyV/329-wip-yuri2-testing-2018-06-19-1953-luminous)

**Comment by yuriw** [comment]
@batrick I suspect this OR is breaking build, pls see 

https://shaman.ceph.com/builds/ceph/wip-yuri2-testing-2018-06-19-1953-luminous/

https://jenkins.ceph.com/job/ceph-dev-new-build/ARCH=x86_64,AVAILABLE_ARCH=x86_64,AVAILABLE_DIST=trusty,DIST=trusty,MACHINE_SIZE=huge/12398//consoleFull

`In file included from /build/ceph-12.2.5-476-gd798583/src/client/Inode.cc:4:0:
/build/ceph-12.2.5-476-gd798583/src/client/Client.h:767:16: warning: 'C_Client_Remount' is already a friend of 'Client' [enabled by default]
   friend class C_Client_Remount;
                ^
/build/ceph-12.2.5-476-gd798583/src/client/Client.cc:4093:6: error: prototype for 'void Client::trim_caps(MetaSession*, uint64_t)' does not match any in class 'Client'
 void Client::trim_caps(MetaSession *s, uint64_t max)
      ^
In file included from /build/ceph-12.2.5-476-gd798583/src/client/Client.cc:83:0:
/build/ceph-12.2.5-476-gd798583/src/client/Client.h:540:8: error: candidate is: void Client::trim_caps(MetaSession*, int)
   void trim_caps(MetaSession *s, int max);
        ^
make[4]: Leaving directory `/build/ceph-12.2.5-476-gd798583/obj-x86_64-linux-gnu'`

**Comment by batrick** [comment]
@yuriw try again

**Pull request merged**